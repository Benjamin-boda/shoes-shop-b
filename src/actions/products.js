import database from "../firebase/firebase";
import firebase from 'firebase/app'

//ADD_PRODUCT

export const addProduct = (product) => ({
    type: "ADD_PRODUCT",
    product
});

export const startAddProduct = (productData) => {
    return (dispatch) => {
        const uid = firebase.auth().currentUser.uid;
        const {
            image = "",
            name = "",
            size = "",
            price = 0,
            boughtAt = 0
        } = productData;
        const product = {image, name, size, price, boughtAt};

        return database.ref(`users/${uid}/products`).push(product).then((ref) => {
            dispatch(addProduct({
                id: ref.key,
                ...product
            }))
        })
    }
}

//REMOVE_PRODUCT

export const removeProduct = ({id} = {}) => ({
    type: "REMOVE_PRODUCT",
    id
});

export const startRemoveProduct = ({id}) => {
    return (dispatch) => {
        const uid = firebase.auth().currentUser.uid;
        return database.ref(`users/${uid}/products/${id}`).remove().then(() => {
            dispatch(removeProduct({id}))
        })
    }
}

//EDIT_PRODUCT

export const editProduct = (id, updates) => ({
    type: "EDIT_PRODUCT",
    id,
    updates
});

export const startEditProduct = (id, updates) => {
    return (dispatch) => {
        const uid = firebase.auth().currentUser.uid;
        return database.ref(`users/${uid}/products/${id}`).update(updates).then(() => {
            dispatch(editProduct(id, updates))
        })
    }
}

//SET_PRODUCTS

export const setProducts = (products) => ({
    type: "SET_PRODUCTS",
    products
})

export const startSetProducts = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/products`).once("value").then((snapshot) => {
            const products = [];

            snapshot.forEach((childSnapshot) => {
                products.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                })
            })
        dispatch(setProducts(products))
        })
    }
}

//REMOVE ALL

export const resetProducts = () => ({
    type: "RESET_PRODUCTS"
});

export const startResetProducts = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;

        return database.ref(`users/${uid}/products`).remove()
            .then(() => {
                dispatch(resetProducts());
            });
    };
};