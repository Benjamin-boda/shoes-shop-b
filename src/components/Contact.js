import React, { useEffect, useState } from "react";

export const Contact = () => {
    const [subject, setSubject] = useState("")
    const [description, setDescription] = useState("")
    const [submit, setSubmit] = useState("")

    const inputChange = (e) => {
        setSubject(e.target.value)
    }

    const textChange = (e) => {
        setDescription(e.target.value)
    }

    useEffect(() => {
        const request = {
            title: subject,
            object: description
        }
        setSubmit(request)
    }, [subject, description])

    const onSubmit = () => {
        console.log(submit)
    }

    return (
        <div className="contact">
            <div className="contact__header">
                <h1 className="contact__title">Contact us</h1>
                <img className="contact__image" src={require("../images/conctact logo.jpg").default}/>
            </div>
            
            <form className="contact__form" onSubmit={onSubmit}>
                <input data-testid="input" className="contact__form__subject" onChange={inputChange} type="text" placeholder="Subject"/>
                <textarea data-testid="textarea" className="contact__form__description" onChange={textChange} maxLength="200" minLength="10" placeholder="Describe your problem in 200 caracters"/>
                <button className="contact__form__button">Submit</button>
            </form>
        </div>
    )
}