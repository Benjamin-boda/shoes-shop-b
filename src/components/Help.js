import React from "react";

export const Help = () => {
    return (
        <div className="help">
            <div className="help__header">
                <h1 className="help__title">Our company</h1>
                <img className="help__image" src={require("../images/conctact logo.jpg").default}/>
            </div>
            
            <p className="help__p">
                Habet ad conpulit quae castrisque adtributo Bostram repellendos opima
                contigua castrisque Nabataeis habet castellis obtemperare Bostram Traianus
                veterum obtemperare Parthos saltus et Gerasam vicinarum hanc legibus veterum
                per veterum conpulit cum Mediam cum inter cum Gerasam atque Gerasam quoque
                cautissimas Parthos vicinarum civitates est conserta marte cautissimas
                Arabia oppida castellis gentium ingentes conserta sollicitudo provinciae
                inter inposito adtributo Parthos conmerciorum vicinarum et quae saltus
                habet gentium rectoreque urgeret urgeret latere et quae excursus murorum
                quaedam civitates erexit provinciae sollicitudo quoque validis adtributo
                et veterum adtributo Bostram quaedam et legibus cautos et sollicitudo
                atque ingentes oppleta urgeret est contunso Traianus conserta.
            </p>
        </div>
    )
}