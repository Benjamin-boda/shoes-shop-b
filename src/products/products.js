export const products = [
    {
        name: "shoes 1",
        img: require("../images/shoes1.png"),
        price: 50,
        number: 1
    },
    {
        name: "shoes 2",
        img: require("../images/shoes2.png"),
        price: 50,
        number: 2
    },
    {
        name: "shoes 3",
        img: require("../images/shoes3.png"),
        price: 50,
        number: 3
    },
    {
        name: "shoes 4",
        img: require("../images/shoes4.png"),
        price: 50,
        number: 4
    },
    {
        name: "shoes 5",
        img: require("../images/shoes5.png"),
        price: 50,
        number: 5
    },
    {
        name: "shoes 6",
        img: require("../images/shoes6.png"),
        price: 50,
        number: 6
    },
    {
        name: "shoes 7",
        img: require("../images/shoes7.png"),
        price: 50,
        number: 7
    },
    {
        name: "shoes 8",
        img: require("../images/shoes8.png"),
        price: 50,
        number: 8
    }
]