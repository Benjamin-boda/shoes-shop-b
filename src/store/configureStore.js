import {createStore, combineReducers, compose, applyMiddleware } from "redux";
import thunk from "redux-thunk";
import {authReducer} from "../reducers/auth";
import {ordersReducer} from "../reducers/orders";
import filterReducer from "../reducers/filters";
import productsReducer from "../reducers/product";

// Store creation

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION__COMPOSE__ || compose;

export default () => {
    const store = createStore(
        combineReducers({
            orders: ordersReducer,
            products: productsReducer,
            filters: filterReducer,
            auth: authReducer
        }),
        composeEnhancers(applyMiddleware(thunk))
        // window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
    );
    return store;
};