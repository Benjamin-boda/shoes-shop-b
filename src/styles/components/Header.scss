.header {
    align-items: center;
    border-bottom: solid grey 1px;
    display: flex;
    font-family: "MonserratRegular";
    height: 15vh;
    justify-content: flex-start;
    margin: 0 auto;
    max-width: 110rem;

    @media (min-width: $desktop-breakpoint-large) {
        justify-content: space-between;
    }
}

.header__logo {
    width: 5rem;

    @media (min-width: $desktop-breakpoint) {
        width: 8rem;
    }
}

.header__bag {
    position: absolute;
    width: 3rem;
    margin-left: $xl-size;

    @media (min-width: $desktop-breakpoint) {
        margin-left: 10rem;
    }

    @media (min-width: $desktop-breakpoint-large) {
        margin-left: 0;
    }
}

.header__link {
    color: $purple;
    font-size: $font-size-xxsmall;
    margin-left: $xs-size;
    text-decoration: none;

    @media (min-width: $desktop-breakpoint) {
        font-size: $font-size-small;
        margin-left: $m-size;
    }

    @media (min-width: $desktop-breakpoint-large) {
        margin-left: 0;
    }
}

.header__link__account {
    color: $purple;
    font-size: $font-size-xxsmall;
    position: absolute;
    text-decoration: none;

    @media (min-width: $desktop-breakpoint) {
        font-size: $font-size-small;
    }
}

.header__link__account__dropdown {
    text-decoration: none;
}

.header__account {
    align-items: center;
    display: flex;
    justify-content: center;
    margin-left: $xl-size;

    @media (min-width: $desktop-breakpoint) {
        margin-left: 8rem;
    }

    @media (min-width: $desktop-breakpoint-large) {
        margin-left: 0;
    }
}

.header__bag__counter {
    background: red;
    border-radius: 50%;
    color: white;
    font-size: 1rem;
    margin-left: $xl-size;
    padding: .2rem;
    position: relative;
    right: -2rem;
    text-align: center;
    top: -2rem;
    width: 2rem;

    @media (min-width: $desktop-breakpoint) {
        margin-left: 10rem;
    }

    @media (min-width: $desktop-breakpoint-large) {
        margin-left: 0;
    }
}


.header__dropdown {
    background: white;
    border: black .1rem solid;
    border-top: none;
    border-bottom-right-radius: 5px;
    border-bottom-left-radius: 5px;
    font-family: "MonserratBold";
    font-size: $font-size-xxsmall;
    padding: $m-size $s-size;
    position: relative;
    top: 6rem;

    @media (min-width: $desktop-breakpoint) {
        font-size: $font-size-small;
        top: 8rem;
    }
}

.header__dropdown__item {
    color: $purple;
    cursor: pointer;
    list-style-type: none;
    text-decoration: none;
}