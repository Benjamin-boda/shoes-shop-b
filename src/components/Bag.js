import React, { useEffect, useState } from "react";
import {useSelector, useDispatch} from "react-redux";
import {startRemoveProduct, startEditProduct, startResetProducts} from "../actions/products";
import {startAddOrder} from "../actions/orders";
import products from "../tests/fixtures/products";

export const Bag = () => {
    const [sizeUpdate, setSizeUpdate] = useState("")
    const [productUpdate, setProductUpdate] = useState("")
    const productsInStore  = useSelector(state => state.products)
    const dispatch = useDispatch()

    const sizes = [37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47]
    

    const sizeEdit = (e) => {
        setSizeUpdate(e.target.value)
    }

    useEffect(() => {
        if (productUpdate !== "" & productsInStore.lenght > 0) {
            const filterProductArray = productsInStore.filter((product) => productUpdate === product.name)
            const productEdit = {
                image : {
                    default : filterProductArray[0].image.default
                },
                name : filterProductArray[0].name,
                size : sizeUpdate,
                price : filterProductArray[0].price,
                boughtAt : filterProductArray[0].boughtAt
            }
            const id = filterProductArray[0].id
            dispatch(startEditProduct(id, productEdit))    
        }
    }, [sizeUpdate])

    const paymentOrder = () => {
        productsInStore.map((product) => {
            product.boughtAt = Date()
            dispatch(startAddOrder(product))})
        dispatch(startResetProducts())              
    }

    const sumPrice = productsInStore.length > 0 ? productsInStore.map(product => product.price).reduce((acc, price) => acc + price) : "0"

    return (
        <div className="bag__content">
            <div className="bag__box">
                {productsInStore.length > 0 ?
                    productsInStore.map((product) => 
                        <div className="bag" key={product.name} onMouseEnter={() => setProductUpdate(product.name)}>
                            <img className="bag__image" src={product.image.default}/>
                            <p data-testid={product.name} className="bag__name">{product.name}</p>
                            <select className="bag__select" name="size" onChange={sizeEdit}>
                                <option value={product.size}>{product.size}</option>
                                {sizes.map((size) => 
                                    <option value={size} key={size}>{size}</option>
                                    )}
                            </select>
                            <p className="bag__name">Price : {product.price} €</p>
                            <button className="bag__button" onClick={() => dispatch(startRemoveProduct(product))}><img className="bag__cross__button" src={require("../images/cross button.png").default}/></button>
                        </div>
                    ) : <p className="bag__empty">Nothing in the bag</p>
                }
            </div>
            
            <p className="bag__price">Total price : {sumPrice} €</p>

            <div className="bag__payment__div">
                <button className="bag__payment" onClick={paymentOrder} disabled={productsInStore.length === 0 ? true : false}>Payment</button>
            </div>
            
        </div>
    )
}