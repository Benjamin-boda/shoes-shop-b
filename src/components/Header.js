import React, {useState ,useContext} from "react";
import {Link} from "react-router-dom";
import {useDispatch, useSelector} from "react-redux";
import {startLogin, startLogout} from "../actions/auth";
import {AuthContext} from "../firebase/Auth";
import {LoginModal} from "./LoginModal";

export const Header = () => {
    const {currentUser} = useContext(AuthContext);
    const [accountHovering, setAccountHovering] = useState(false)
    const [modalIsOpen, setModalIsOpen] = useState(false)
    const productsInStore = useSelector(state => state.products)
    const dispatch = useDispatch()

    const openModal = () => {
        if (!!currentUser) {
            setModalIsOpen(false)
        } else {
            setModalIsOpen(true)
        }
    }

    return (
        <div className="header">
            <img className="header__logo" src={require("../images/logo.png").default}/>
            <Link className="header__link" to="/">Home</Link>
            <Link className="header__link" to={!!currentUser ? "/shop/1" : ""} onClick={openModal}>Shop</Link>
            <Link className="header__link" to="/contact">Contact</Link>
            <Link className="header__link" to="/help">Help</Link>
            
            <div className="header__account">
                <Link 
                    data-testid="account"
                    className="header__link__account"
                    to=""
                    onClick={() => !!currentUser ? undefined : dispatch(startLogin())}
                    onMouseEnter={() => !!currentUser ? setAccountHovering(true) : setAccountHovering(false)}
                >
                    {!!currentUser ? "User account" : "Login"}
                </Link>
                {accountHovering ? 
                    <ul className="header__dropdown" onClick={() => setAccountHovering(false)} onMouseLeave={() => setAccountHovering(false)}>
                        <Link className="header__link__account__dropdown" to="/orders">
                            <li className="header__dropdown__item">My account</li>
                        </Link>
                        <li className="header__dropdown__item" onClick={() => dispatch(startLogout())}>Logout</li>
                        <li className="header__dropdown__item" onClick={() => setAccountHovering(false)}>Close</li>
                    </ul> : undefined
                }
            </div>
            
            <div>
                <Link to={!!currentUser ? "/bag" : ""} onClick={openModal}><img className="header__bag" src={require("../images/bag.png").default}/></Link>
                <p className="header__bag__counter">{productsInStore.length > 0 ? productsInStore.length : 0}</p>
            </div>
            

            <LoginModal modalIsOpen={modalIsOpen} setModalIsOpen={setModalIsOpen}/>
        </div>
    )
}