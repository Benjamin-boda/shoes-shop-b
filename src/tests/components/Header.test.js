import React from "react";
import { useSelector } from 'react-redux';
import configureStore from "redux-mock-store";
import {Header} from "../../components/Header";
import {AuthProvider} from "../../firebase/Auth";
import {render, cleanup, screen} from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect';
import products from "../fixtures/products";
import {productsReducer} from "../../reducers/product";
import { MemoryRouter } from 'react-router-dom';

const initialState = {
    state : {
        products: products
    }
};


jest.mock('react-redux', () => ({
    useDispatch: jest.fn(),
    useSelector: jest.fn()
}));

useSelector.mockImplementation((selector) => selector({
    products : products
}))

export const renderWithRedux = (ui, { initialState, ...renderOptions } = {}) => {
    const mockStore = configureStore(productsReducer, initialState);
    const store = mockStore({});
    const Wrapper = ({ children }) => (
        <MemoryRouter>
            <AuthProvider store={store}>{children}</AuthProvider>
        </MemoryRouter>
    );
    return render(ui, { wrapper: Wrapper, ...renderOptions });
  };

const { asFragment, getByTestId } = renderWithRedux(<Header />, initialState)


afterEach(cleanup)

test("should match the snapshot", () => {
    
    expect(asFragment(<Header/>)).toMatchSnapshot()
})

test("user not logged, Login must appear", () => {
    useSelector.mockImplementation((selector) => selector({
        products : products
    }))
    renderWithRedux(<Header />, initialState)
    expect(screen.getByTestId("account")).toHaveTextContent("Login")
})