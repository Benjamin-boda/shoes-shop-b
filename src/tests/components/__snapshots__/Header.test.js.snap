// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="header"
  >
    <img
      class="header__logo"
    />
    <a
      class="header__link"
      href="/"
    >
      Home
    </a>
    <a
      class="header__link"
      href="/"
    >
      Shop
    </a>
    <a
      class="header__link"
      href="/contact"
    >
      Contact
    </a>
    <a
      class="header__link"
      href="/help"
    >
      Help
    </a>
    <div
      class="header__account"
    >
      <a
        class="header__link__account"
        data-testid="account"
        href="/"
      >
        Login
      </a>
    </div>
    <div>
      <a
        href="/"
      >
        <img
          class="header__bag"
        />
      </a>
      <p
        class="header__bag__counter"
      >
        3
      </p>
    </div>
  </div>
</DocumentFragment>
`;
